FlexMeter.c

```diff
@@ -0,0 +1,226 @@
+
+typedef struct {
+   char* name;
+   char* command;
+   char* type;
+   char* caption;
+   char* uiName;
+}_flex_meter;
+
+_flex_meter meter_list[30];
+
+int meter_list_idx=0;
+static int meters_count=0;
+
+static const int DateMeter_attributes[] = {
+   FLEX
+};
+
+MeterClass * FlexMeter_class = NULL;
+
+int check_for_meters(void);
+
+static int parse_input(char * line)
+{
+   for (int i = strlen(line);i>0;i--)
+   {
+      if ((line[i]=='\r') || (line[i]=='\n'))
+      {
+         line[i]='\0';
+      }
+   }
+
+   if (!strncmp(line,"name=",5))
+   {
+      meter_list[meters_count].uiName = xStrdup(line+5);
+   }
+   else if (!strncmp(line,"command=",7))
+   {
+      meter_list[meters_count].command = xStrdup(line+8);
+   }
+   else if (!strncmp(line,"caption=",7))
+   {
+      meter_list[meters_count].caption = xStrdup(line+8);
+   }
+   else if (!strncmp(line,"type=",5))
+   {
+      meter_list[meters_count].type = xStrdup(line+6);
+   }
+   else
+   {
+      return -1;
+   }
+
+   return 0;
+}
+
+static int load_config(char * file)
+{
+   int ret=0;
+   char buff[2048];
+   memset(buff,0,2048);
+   FILE *fp = fopen( file , "r");
+   while( NULL != fgets(buff, 2048, fp) )
+   {
+      parse_input(buff);
+      memset(buff,0,2048);
+   }
+
+   fclose(fp);
+   return ret;
+}
+
+int check_for_meters(void)
+{
+   char path[400]; // full path
+   DIR *d;
+   struct dirent *dir;
+   struct passwd *pw = getpwuid(getuid());
+   const char *homedir = pw->pw_dir;
+
+   // path to home folder 1 for zero 1 for slash
+   char * home = (char * ) xCalloc(1,(strlen(homedir) + strlen(FLEX_CFG_FOLDER) + 2));
+
+   sprintf(home,"%s/%s",homedir,FLEX_CFG_FOLDER);
+
+   d = opendir(home);
+   if (d) {
+      while ((dir = readdir(d)) != NULL) {
+         if ( dir->d_name[0]!='.' && strcmp(".",dir->d_name) && strcmp("..",dir->d_name))
+         {
+            meter_list[meters_count].name = xStrdup(dir->d_name);
+            memset(path,0,80);
+            sprintf(path,"%s/%s",home,dir->d_name);
+            if( 0 != load_config(path) )
+            {
+               break;
+            }
+
+            if ( meters_count < MAX_METERS_COUNT )
+            {
+               meters_count++;
+            }
+            else
+            {
+               break; // go out we reach the limit
+            }
+         }
+      }
+      closedir(d);
+   }
+   else
+   {
+      if (0 == mkdir(home,0777))
+      {
+         char template [4][40] =
+         {
+            "name=template",
+            "command=echo \"`uptime`\"",
+            "type=TEXT_METERMODE",
+            "caption=\"UPTIME\""
+         };
+
+         char  template_filename[500];
+         sprintf(template_filename,"%s/%s",home,".Template");
+         FILE * fp = fopen(template_filename,"w");
+         if (fp != NULL )
+         {
+            for (int i=0; i<4; i++)
+            {
+               fprintf(fp,"%s\n",template[i]);
+            }
+         }
```

Comment:
I suggest not to create any sample file here. When the directory doesn't exist, it is likely that user does not want any flex meter at all. Besides, creating a file that runs an external command is always risky. You can't tell whether the `uptime` program from user's device is always a safe one. If that had been compromised already, you would aid an attack.

Reply:
It was more for convenience this part, I can drop it and just document it or even place .Template file in `doc` folder of the project

Reply:
@bogdanovs I'm okay with sample files. But just don't create them in the user's folders without their consent. They can read the manual and make one if they want. And yes, go ahead and put a sample in the `doc` folder of this source repo.

Reply:
Why not use `const char* template[] = {…, NULL}` for the definition?